In [1]:
import numpy as np
import pandas as pd

In [4]:
data = pd.read_csv('laptopData.csv')
df = data.copy()
df1 = data.copy()

In [112]:
data.head()
data[data['Weight']=='?']

,Unnamed: 0,Company,TypeName,Inches,ScreenResolution,Cpu,Ram,Memory,Gpu,OpSys,Weight,Price
208,208.0,Dell,Ultrabook,13.3,Full HD 1920x1080,Intel Core i7 8550U 1.8GHz,8GB,256GB SSD,Intel UHD Graphics 620,Windows 10,?,77202.72


#### Summary of data
The Uncleaned Laptop Price dataset is a collection of laptop product listings scraped from an online e-commerce website. The dataset includes information about various laptop models, such as their brand, screen size, processor, memory, storage capacity, operating system, and price. However, the dataset is uncleaned, meaning that it contains missing values, inconsistent formatting, and other errors that need to be addressed before the data can be used for analysis or modeling.

The dataset contains both categorical and numerical variables, with the majority of variables being categorical, including brand, model name, screen resolution, processor type, and operating system. Some numerical variables include screen size, memory, and storage capacity. The target variable in the dataset is the price, which is a continuous variable.

The dataset contains over 1,300 laptop listings, making it a reasonably sized dataset for analysis and modeling. The dataset may be useful for machine learning projects related to predicting the price of a laptop based on its specifications. However, before using the dataset, it would be necessary to clean and preprocess the data to address the inconsistencies and missing values.

### Issues with the Dataset

#### Dirty Data
##### 0. Company and typeName
- Rows:
  - 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245        `completeness`
##### 1. Inches
- Misssing values: 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245 `completeness`
- Abnormal inches:
  - 408, 423, 673, 438, 525, 402, 424, 458, 592 `accuracy`
- Row 476 has '?'  `validation`
- str type should be it float `validation`
##### 2. Ram
- Misssing values: 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245     `completeness`
- 'GB' is extra part `accuracy`
- int str `validation`
##### 3. Memory
- Misssing values: 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245     `completeness`
- 1.0TB Hybrid `accuracy`
##### 4. Weight
- Misssing values: 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245     `completeness`
- Abnormal weight:
  - 133, 173, 208, 302, 326, 349, 587, 656  `accuracy`
- Row 208 has '?'  `validation`
- 'Kg' extra  `validation`
- in str convert in float  `validation`
##### 5. Price
- Misssing values: 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245     `completeness`
- float and int both data   `validation`
- some float values have 4 decimal digits and some 2 or 3 `accuracy`
##### 6. Unnamed: 0
- Misssing values: 20, 46, 50, 115, 125, 148, 190, 209, 267, 336, 350, 478, 567, 616, 625, 629, 666, 699, 726, 747, 753, 774, 922, 923, 938, 978, 994, 1060, 1202, 1245     `completeness`
- is float type  `validation`

---

#### Messy Data

##### 1. ScreenResolution Column
- can be split into 3 cols:
  - Resolution
  - Touchscreen
  - Feature
##### 2. Memory Column
- can be split in:
  - Rom
  - Type
##### 3. Cpu Column
- can be split into 2 cols:
  - Cpu
  - Cooling Speed

In [7]:
df.drop_duplicates(inplace=True)

In [8]:
df.drop(20,inplace=True)

In [9]:
df.drop(columns='Unnamed: 0',inplace=True)

In [19]:
df['Inches'] = df['Inches'].str.replace('?','15.6')

In [21]:
df['Inches'] = df['Inches'].astype('float64')

In [ ]:
df['Ram'] = df['Ram'].str.replace('GB','')

In [12]:
df['Ram'] = df['Ram'].astype('int32')

In [54]:
df1['Memory'] = df['Memory'].str.replace('.0TB', 'TB', regex=False)
df1['Memory'] = df['Memory'].str.replace('TB', '000GB', regex=False)

In [72]:
df['SSD'] = df['Memory'].str.extract(r'(\d+)GB SSD')[0].fillna(0).astype(int)

df['HDD'] = df['Memory'].str.extract(r'(\d+)GB HDD')[0].fillna(0).astype(int)

df['Hybrid'] = df['Memory'].str.extract(r'(\d+)GB Hybrid')[0].fillna(0).astype(int)

df['Flash_Storage'] = df['Memory'].str.extract(r'(\d+)GB Flash Storage')[0].fillna(0).astype(int)

In [87]:
df.drop(columns='Memory',inplace=True)

In [129]:
df['Weight'] = df['Weight'].str.replace('kg','')
df['Weight'] = df['Weight'].replace('', np.nan)
df['Weight'] = pd.to_numeric(df['Weight'])

In [132]:
df['Weight'] = df['Weight'].astype('float64')

In [ ]:
df['Weight'] = df['Weight'].fillna(df['Weight'].median(), inplace=True)

<class 'pandas.DataFrame'>
Index: 1273 entries, 0 to 1302
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1273 non-null   str    
 1   TypeName          1273 non-null   str    
 2   Inches            1273 non-null   float64
 3   ScreenResolution  1273 non-null   str    
 4   Cpu               1273 non-null   str    
 5   Ram               1273 non-null   int32  
 6   Gpu               1273 non-null   str    
 7   OpSys             1273 non-null   str    
 8   Weight            1273 non-null   float64
 9   Price             1273 non-null   float64
 10  SSD               1273 non-null   int64  
 11  HDD               1273 non-null   int64  
 12  Hybrid            1273 non-null   int64  
 13  Flash_Storage     1273 non-null   int64  
dtypes: float64(3), int32(1), int64(4), str(6)
memory usage: 176.5 KB


C:\Users\moham\AppData\Local\Temp\ipykernel_17556\1208914148.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df['Weight'] = df['Weight'].fillna(df['Weight'].median(), inplace=True)


In [160]:
df1.dropna(inplace=True)

In [170]:
df['Touchscreen'] = df['ScreenResolution'].apply(lambda x: 1 if 'Touchscreen' in x else 0)
df['Resolution'] = df['ScreenResolution'].str.extract(r'(\d+x\d+)')
df['Feature'] = df['ScreenResolution'].str.replace(r'\d+x\d+', '', regex=True).str.strip()
df['Feature'] = df['Feature'].str.replace('Touchscreen', '', regex=False).str.replace('/', '').str.strip()
df['Feature'] = df['Feature'].replace('', 'Normal')


In [181]:
df['Cpu_Speed'] = df['Cpu'].str.extract(r'(\d+\.?\d*)GHz').astype(float)
df['Cpu_Name'] = df['Cpu'].str.replace(r'\d+\.?\d*GHz', '', regex=True).str.strip()
df.drop(columns='Cpu',inplace=True)

In [184]:
# pd.set_option('display.max_rows',None)
# pd.reset_option('display.max_rows')

df

,Company,TypeName,Inches,ScreenResolution,Ram,Gpu,OpSys,Weight,Price,SSD,HDD,Hybrid,Flash_Storage,Resolution,Touchscreen,Feature,Cpu_Speed,Cpu_Name
0,Apple,Ultrabook,13.3,IPS Panel Retina Display 2560x1600,8,Intel Iris Plus Graphics 640,macOS,1.37,71378.6832,128,0,0,0,2560x1600,0,IPS Panel Retina Display,2.3,Intel Core i5
1,Apple,Ultrabook,13.3,1440x900,8,Intel HD Graphics 6000,macOS,1.34,47895.5232,0,0,0,128,1440x900,0,Normal,1.8,Intel Core i5
2,HP,Notebook,15.6,Full HD 1920x1080,8,Intel HD Graphics 620,No OS,1.86,30636.0000,256,0,0,0,1920x1080,0,Full HD,2.5,Intel Core i5 7200U
3,Apple,Ultrabook,15.4,IPS Panel Retina Display 2880x1800,16,AMD Radeon Pro 455,macOS,1.83,135195.3360,512,0,0,0,2880x1800,0,IPS Panel Retina Display,2.7,Intel Core i7
4,Apple,Ultrabook,13.3,IPS Panel Retina Display 2560x1600,8,Intel Iris Plus Graphics 650,macOS,1.37,96095.8080,256,0,0,0,2560x1600,0,IPS Panel Retina Display,3.1,Intel Core i5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1298,Lenovo,2 in 1 Convertible,14.0,IPS Panel Full HD / Touchscreen 1920x1080,4,Intel HD Graphics 520,Windows 10,1.80,33992.6400,128,0,0,0,1920x1080,1,IPS Panel Full HD,2.5,Intel Core i7 6500U
1299,Lenovo,2 in 1 Convertible,13.3,IPS Panel Quad HD+ / Touchscreen 3200x1800,16,Intel HD Graphics 520,Windows 10,1.30,79866.7200,512,0,0,0,3200x1800,1,IPS Panel Quad HD+,2.5,Intel Core i7 6500U
1300,Lenovo,Notebook,14.0,1366x768,2,Intel HD Graphics,Windows 10,1.50,12201.1200,0,0,0,64,1366x768,0,Normal,1.6,Intel Celeron Dual Core N3050
1301,HP,Notebook,15.6,1366x768,6,AMD Radeon R5 M330,Windows 10,2.19,40705.9200,0,1000,0,0,1366x768,0,Normal,2.5,Intel Core i7 6500U


In [187]:
df.info()

<class 'pandas.DataFrame'>
Index: 1273 entries, 0 to 1302
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1273 non-null   str    
 1   TypeName          1273 non-null   str    
 2   Inches            1273 non-null   float64
 3   ScreenResolution  1273 non-null   str    
 4   Ram               1273 non-null   int32  
 5   Gpu               1273 non-null   str    
 6   OpSys             1273 non-null   str    
 7   Weight            1273 non-null   float64
 8   Price             1273 non-null   float64
 9   SSD               1273 non-null   int64  
 10  HDD               1273 non-null   int64  
 11  Hybrid            1273 non-null   int64  
 12  Flash_Storage     1273 non-null   int64  
 13  Resolution        1273 non-null   str    
 14  Touchscreen       1273 non-null   int64  
 15  Feature           1273 non-null   str    
 16  Cpu_Speed         1273 non-null   float64
 17  Cpu_Name   

In [189]:
df.describe()

,Inches,Ram,Weight,Price,SSD,HDD,Hybrid,Flash_Storage,Touchscreen,Cpu_Speed
count,1273.000000,1273.000000,1273.000000,1273.000000,1273.000000,1273.000000,1273.000000,1273.000000,1273.000000,1273.000000
mean,15.131186,8.462687,2.077589,59955.814073,182.309505,414.108405,9.040063,4.587588,0.145326,2.298916
std,1.953712,5.564408,0.807491,37332.251005,184.305619,515.266937,93.642364,30.566461,0.352568,0.507118
min,10.100000,1.000000,0.000200,9270.720000,0.000000,0.000000,0.000000,0.000000,0.000000,0.900000
25%,14.000000,4.000000,1.500000,31914.720000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000
50%,15.600000,8.000000,2.040000,52161.120000,256.000000,0.000000,0.000000,0.000000,0.000000,2.500000
75%,15.600000,8.000000,2.320000,79333.387200,256.000000,1000.000000,0.000000,0.000000,0.000000,2.700000
max,35.600000,64.000000,11.100000,324954.720000,1000.000000,2000.000000,1000.000000,512.000000,1.000000,3.600000


In [186]:
df.to_csv('Cleaned_LaptopData.csv',index=False)